### PR TITLE
Find parallel NetCDF with cmake for USER-NETCDF

### PR DESCRIPTION
**Pull request by jotelha**
**Summary**

USER-NETCDF module can be built with NetCDF or Parallel NetCDF library, or both. Depending on which one is available, different dump styles are compiled, namely "netcdf" for the former, "netcdf/mpiio" for the latter. So far, the new CMAKE build system fails to find and compile with Parallel NetCDF. An additional FindPNetCDF.cmake snippet in combination with some extensions within USER-NETCDF.cmake allow LAMMPS with USER-NETCDF to build both styles, depending on which library is available. 

**Author(s)**

Johannes Hörmann, johannes.hoermann@imtek.uni-freiburg.de, johannes.laurin@gmail.com
FindPNetCDF.cmake snippet by unknown author from https://ftp.space.dtu.dk/pub/Ioana/pism0.6.1-10/CMake/FindPNetCDF.cmake, http://www.pism-docs.org

_Please state name and affiliation of the author or authors that should be credited with the changes in this pull request. If this pull request adds new files to the distribution, please also provide a suitable "long-lived" e-mail address (ideally something that can outlive your institution's e-mail, in case you change jobs) for the *corresponding* author, i.e. the person the LAMMPS developers can contact directly with questions and requests related to maintenance and support of this contributed code._

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

Note:  The FindPNetCDF.cmake resides under GPL v3 at https://ftp.space.dtu.dk/pub/Ioana/pism0.6.1-10/CMake/FindPNetCDF.cmake and I am not sure whether this would conflict with LAMMPS GPL v2. 

**Backward Compatibility**

**Implementation Notes**

Correctness was tested with various custom test cases for netcdf dump styles at https://github.com/jotelha/lammps-testing/tree/master/tests/examples/USER/netcdf

_Provide any relevant details about how the changes are implemented, how correctness was verified, how other features - if any - in LAMMPS are affected_

**Post Submission Checklist**

_Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply_

- [*] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [*] The feature has been verified to work with the CMake based build system
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

_Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)_

### CODE REVIEW

File: cmake/Modules/FindNetCDF.cmake [hunk 1]
```diff
@@ -46,10 +46,14 @@ endif()
...
 mark_as_advanced (NETCDF_INCLUDE_DIR)
+
 set (NETCDF_C_INCLUDE_DIRS ${NETCDF_INCLUDE_DIR})
 
+string(REGEX REPLACE "/include/?$" "/lib"
+  NETCDF_LIB_HINT ${NETCDF_INCLUDE_DIR})
+
 find_library (NETCDF_LIBRARY NAMES netcdf
-  HINTS "${NETCDF_DIR}/lib")
+  HINTS "${NETCDF_DIR}/lib" "${NETCDF_LIB_HINT}")
```
Comment (junghans) [inline cmake/Modules/FindNetCDF.cmake]:
```suggestion
  HINTS "${NETCDF_DIR}" "${NETCDF_LIB_HINT}" PATH_SUFFIXES lib lib64)
```

File: cmake/Modules/FindNetCDF.cmake [hunk 2]
```diff
@@ -46,10 +46,14 @@ endif()
 find_path (NETCDF_INCLUDE_DIR netcdf.h
   HINTS "${NETCDF_DIR}/include")
 mark_as_advanced (NETCDF_INCLUDE_DIR)
+
 set (NETCDF_C_INCLUDE_DIRS ${NETCDF_INCLUDE_DIR})
 
+string(REGEX REPLACE "/include/?$" "/lib"
```
Comment (junghans) [inline cmake/Modules/FindNetCDF.cmake]:
```suggestion
string(REGEX REPLACE "/include/?$" ""
```

File: cmake/Modules/FindPNetCDF.cmake [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+  file(TO_CMAKE_PATH "$ENV{LD_LIBRARY_PATH}" LD_LIBRARY_PATH)
+
+  find_library(PNETCDF_LIBRARIES
+    NAMES pnetcdf
+    HINTS ${LD_LIBRARY_PATH})
+
+  if (PNETCDF_LIBRARIES)
+    get_filename_component(PNETCDF_LIB_DIR ${PNETCDF_LIBRARIES} PATH)
+    string(REGEX REPLACE "/lib/?$" "/include"
```
Comment (junghans) [inline cmake/Modules/FindPNetCDF.cmake]:
```suggestion
    string(REGEX REPLACE "/(lib|lib64)/?$" "/include"
```

**Comment by akohlmey** [comment]
@junghans @rbberger any more comments or requests? otherwise please approve. thanks.

File: cmake/Modules/FindPNetCDF.cmake [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+if (PNETCDF_INCLUDES AND PNETCDF_LIBRARIES)
+  # Already in cache, be silent
+  set (PNETCDF_FIND_QUIETLY TRUE)
+endif (PNETCDF_INCLUDES AND PNETCDF_LIBRARIES)
+
+find_path (PNETCDF_INCLUDES pnetcdf.h
+  HINTS "${PNETCDF_ROOT}/include" "$ENV{PNETCDF_ROOT}/include")
+
+string(REGEX REPLACE "/include/?$" "/lib"
```
Comment (junghans) [inline cmake/Modules/FindPNetCDF.cmake]:
```suggestion
string(REGEX REPLACE "/include/?$" ""
```

File: cmake/Modules/FindPNetCDF.cmake [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  PNETCDF_LIB_HINT ${PNETCDF_INCLUDES})
+
+find_library (PNETCDF_LIBRARIES
+  NAMES pnetcdf
+  HINTS ${PNETCDF_LIB_HINT})
```
Comment (junghans) [inline cmake/Modules/FindPNetCDF.cmake]:
```suggestion
  HINTS ${PNETCDF_LIB_HINT} PATH_SUFFIXES lib lib64)
```

**Pull request merged**